File: lib/state/tailscale_controller.dart
import 'package:flutter/foundation.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:luci_mobile/main.dart';
import 'package:luci_mobile/models/tailscale.dart';

/// Owns the Tailscale module's state. Kept separate from the large [AppState]
/// so the feature is self-contained; it borrows the authenticated RPC channel
/// via [AppState.rpcCall].
final tailscaleControllerProvider = ChangeNotifierProvider<TailscaleController>(
  (ref) => TailscaleController(ref),
);

class TailscaleController extends ChangeNotifier {
  TailscaleController(this._ref);

  final Ref _ref;

  TailscaleStatus _status = TailscaleStatus.empty;
  TailscaleSettings _settings = TailscaleSettings.empty;
  bool _loaded = false;
  bool _isLoading = false;
  bool _isBusy = false;
  String? _error;

  TailscaleStatus get status => _status;
  TailscaleSettings get settings => _settings;
  bool get loaded => _loaded;
  bool get isLoading => _isLoading;

  /// A write (set_settings) is in flight — used to disable toggles.
  bool get isBusy => _isBusy;
  String? get error => _error;

  Future<dynamic> _rpc(
    String object,
    String method, [
    Map<String, dynamic>? params,
  ]) {
    return _ref
        .read(appStateProvider)
        .rpcCall(object: object, method: method, params: params);
  }

  Future<void> load() async {
    _isLoading = true;
    _error = null;
    notifyListeners();
    try {
      final results = await Future.wait([
        _rpc('tailscale', 'get_status'),
        _rpc('tailscale', 'get_settings'),
      ]);
      final statusData = results[0];
      final settingsData = results[1];
      if (statusData is Map) {
        _status = TailscaleStatus.fromJson(
          Map<String, dynamic>.from(statusData),
        );
      }
      if (settingsData is Map) {
        _settings = TailscaleSettings.fromJson(
          Map<String, dynamic>.from(settingsData),
        );
      }
      _loaded = true;
    } catch (e) {
      _error = e.toString();
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  String get _currentExitNodeIp => _status.currentExitNode?.ip ?? '';

  Future<bool> _apply(Map<String, dynamic> formData) async {
    _isBusy = true;
    _error = null;
    notifyListeners();
    try {
      final res = await _rpc('tailscale', 'set_settings', {
        'form_data': formData,
      });
      if (res is Map && res['error'] != null) {
        _error = res['error'].toString();
        notifyListeners();
        return false;
      }
      await load();
      return true;
    } catch (e) {
      _error = e.toString();
      notifyListeners();
      return false;
    } finally {
      _isBusy = false;
      notifyListeners();
    }
  }

  /// Flip a single boolean setting (e.g. `accept_routes`, `shields_up`,
  /// `disable_magic_dns`, `advertise_exit_node`), preserving everything else.
  Future<bool> setFlag(String key, bool value) {
    final data = _settings.toFormData(
      exitNodeIp: _currentExitNodeIp,
      overrides: {key: value ? '1' : '0'},
    );
    return _apply(data);
  }

  /// Select an exit node by IP, or pass null/'' to clear it.
  Future<bool> setExitNode(String? ip) {
    final target = ip ?? '';
    final data = _settings.toFormData(
      exitNodeIp: target,
      overrides: {'exit_node': target},
    );
    return _apply(data);
  }
}

File: native/Lucinate/Features/Tailscale/TailscaleController.swift
import Foundation
import Observation

/// Owns the Tailscale module's state, mirroring
/// `lib/state/tailscale_controller.dart`. The feature stays self-contained:
/// the authenticated RPC channel is borrowed per-call via the `RouterService`
/// passed into each method, so the controller never retains `AppState`.
@MainActor
@Observable
final class TailscaleController {
    private(set) var status: TailscaleStatus = .empty
    private(set) var settings: TailscaleSettings = .empty
    private(set) var loaded = false
    private(set) var isLoading = false

    /// A write (`set_settings`) is in flight — used to disable toggles.
    private(set) var isBusy = false
    private(set) var error: String?

    /// The `tailscale` rpcd object is absent (plugin not installed on the
    /// router) or the daemon reports `not_installed`. Drives the empty state
    /// instead of an error — this is expected on most routers.
    private(set) var notInstalled = false

    /// Forget everything (used when the selected router changes).
    func reset() {
        status = .empty
        settings = .empty
        loaded = false
        isLoading = false
        isBusy = false
        error = nil
        notInstalled = false
    }

    func load(service: RouterService) async {
        isLoading = true
        error = nil
        defer { isLoading = false }
        do {
            async let statusCall = service.tailscaleStatus()
            async let settingsCall = service.tailscaleSettings()
            let (statusJSON, settingsJSON) = try await (statusCall, settingsCall)
            status = TailscaleStatus.fromJSON(statusJSON)
            settings = TailscaleSettings.fromJSON(settingsJSON)
            notInstalled = !status.isInstalled
            loaded = true
        } catch let ubusError as UbusError where ubusError.isUnavailableObject {
            // The optional tailscale rpcd plugin isn't there — degrade to the
            // "not installed" empty state rather than an error.
            notInstalled = true
            loaded = true
        } catch is CancellationError {
            // View went away mid-load; keep whatever state we had.
        } catch {
            if !Task.isCancelled {
                self.error = error.localizedDescription
            }
        }
    }

    private func apply(formData: [String: JSONValue], service: RouterService) async -> Bool {
        isBusy = true
        error = nil
        defer { isBusy = false }
        do {
            let result = try await service.tailscaleApply(formData: formData)
            if !result["error"].isNull {
                error = result["error"].coercedString ?? "Action failed"
                return false
            }
            await load(service: service)
            return true
        } catch {
            self.error = error.localizedDescription
            return false
        }
    }

    /// Flip a single boolean setting (e.g. `accept_routes`, `shields_up`,
    /// `disable_magic_dns`, `advertise_exit_node`), preserving everything else
    /// — `set_settings` requires the FULL form, so unspecified flags would be
    /// cleared otherwise.
    @discardableResult
    func setFlag(_ key: String, value: Bool, service: RouterService) async -> Bool {
        let data = settings.toFormData(
            exitNodeIp: status.currentExitNode?.ip ?? "",
            overrides: [key: .string(value ? "1" : "0")]
        )
        return await apply(formData: data, service: service)
    }

    /// Select an exit node by IP, or pass nil/"" to clear it.
    @discardableResult
    func setExitNode(ip: String?, service: RouterService) async -> Bool {
        let target = ip ?? ""
        let data = settings.toFormData(
            exitNodeIp: target,
            overrides: ["exit_node": .string(target)]
        )
        return await apply(formData: data, service: service)
    }
}
